File: SegmentTrees/segment_tree.js
/**
 * SegmentTree class for efficient range queries and updates.
 *
 * @param {Array} data - Input array to build the segment tree.
 * @param {Function} func - Associative function (e.g., Math.min, Math.max, sum).
 * @param {*} identity - Identity value for the associative function.
 *                       For Math.min: Infinity, Math.max: -Infinity, sum: 0.
 */
class SegmentTree {
    constructor(data, func = Math.min, identity = null) {
        this.n = data.length; // Length of the input array
        this.func = func; // Associative function (e.g., min, max, sum)

        // Determine the identity value based on the function or use a custom one
        this.identity = identity !== undefined && identity !== null ? identity :
            (func === Math.min ? Infinity :
                func === Math.max ? -Infinity :
                    func === ((a, b) => a + b) ? 0 : null);

        // Throw an error if no valid identity value is provided for custom functions
        if (this.identity === null) {
            throw new Error("Invalid arguments: identity is required for custom functions or unrecognized func.");
        }

        // Create a flat array to represent the segment tree with a size of 2 * n
        this.tree = new Array(this.n * 2);
        this.tree.fill(this.identity); // Initialize with the identity value

        // Build the segment tree from the input data
        this._build(data);
    }

    /**
     * Internal method to build the segment tree from the input array.
     *
     * @param {Array} data - The input array.
     */
    _build(data) {
        // Initialize the leaves (second half of the tree array)
        for (let i = 0; i < this.n; i++) {
            this.tree[this.n + i] = data[i]; // Store data elements as leaves
        }

        // Compute internal nodes (bottom-up approach)
        for (let i = this.n - 1; i > 0; i--) {
            this.tree[i] = this.func(this.tree[2 * i], this.tree[2 * i + 1]); // Combine child nodes
        }
    }

    /**
     * Update the value at a specific index in the original array.
     *
     * @param {number} index - The index in the original array (0-based).
     * @param {*} value - The new value to set.
     */
    update(index, value) {
        // Ensure the index is within valid bounds
        if (index < 0 || index >= this.n) {
            throw new Error("Index out of bounds");
        }

        // Update the corresponding leaf node
        index += this.n; // Map the original index to its corresponding leaf node
        this.tree[index] = value;

        // Propagate changes up the tree to update internal nodes
        while (index > 1) {
            index = Math.floor(index / 2); // Move to the parent node
            this.tree[index] = this.func(this.tree[2 * index], this.tree[2 * index + 1]); // Update the parent
        }
    }

    /**
     * Query the result of the function over a range [left, right).
     *
     * @param {number} left - Start of the range (inclusive, 0-based).
     * @param {number} right - End of the range (exclusive, 0-based).
     * @return {*} - The result of the associative function over the range.
     */
    query(left, right) {
        // Ensure the range is valid
        if (left < 0 || right > this.n || left >= right) {
            throw new Error("IndexError: Invalid query range.");
        }

        // Map the range to corresponding leaf nodes
        left += this.n;
        right += this.n;

        let result = this.identity; // Start with the identity value

        // Traverse and combine results for overlapping segments
        while (left < right) {
            if (left % 2 === 1) { // If left is a right child
                result = this.func(result, this.tree[left]); // Include this segment
                left++; // Move to the next segment
            }
            if (right % 2 === 1) { // If right is a right child
                right--; // Move to the previous segment
                result = this.func(result, this.tree[right]); // Include this segment
            }

            // Move to the parent level
            left = Math.floor(left / 2);
            right = Math.floor(right / 2);
        }

        return result; // Return the combined result
    }
}

// Export the SegmentTree class for use in other files
module.exports = SegmentTree;

File: SegmentTrees/segment_tree.py
class SegmentTree:
    def __init__(self, data, func=min, identity=None):
        """
        Initialize the Segment Tree.

        :param data: The input array for which the segment tree is constructed.
        :param func: The function to apply to a range (e.g., min, max, sum). Must be associative.
        :param identity: The identity value for the given function. If not provided, it will be inferred for common functions.

        The segment tree will store results of `func` for overlapping ranges, allowing fast queries and updates.
        """
        self.n = len(data)  # Number of elements in the original array
        self.func = func    # Function to apply on segments

        # Determine the identity value (default neutral value for the given function)
        self.identity = identity if identity is not None else (
            float('inf') if func == min else
            float('-inf') if func == max else
            0 if func == sum else
            None
        )
        if self.identity is None:
            raise ValueError("Identity value required for custom functions")

        # The segment tree is represented as a flat array of size 2 * n
        self.tree = [self.identity] * (2 * self.n)

        # Build the segment tree from the input data
        self._build(data)

        print(self.tree)

    def _build(self, data):
        """
        Build the Segment Tree from the input data.
        The input data is stored as leaves in the tree, and internal nodes are computed based on `func`.
        """
        # Initialize the leaves (second half of the tree array)
        for i in range(self.n):
            self.tree[self.n + i] = data[i]

        # Compute internal nodes (bottom-up approach)
        for i in range(self.n - 1, 0, -1):  # Start from last internal node
            self.tree[i] = self.func(self.tree[2 * i], self.tree[2 * i + 1])

    def update(self, index, value):
        """
        Update the value at a specific index in the original array.
        Updates the leaf node and propagates changes to internal nodes.

        :param index: The index in the original array (0-based).
        :param value: The new value to set.
        """
        if not (0 <= index < self.n):
            raise IndexError("Index out of bounds")

        # Update the leaf node
        index += self.n  # Map the original index to its corresponding leaf node in the tree
        self.tree[index] = value

        # Update all affected internal nodes
        while index > 1:
            index //= 2  # Move to the parent node
            self.tree[index] = self.func(self.tree[2 * index], self.tree[2 * index + 1])

    def query(self, left, right):
        """
        Query the result of the function in the range [left, right).

        :param left: The starting index of the range (inclusive, 0-based).
        :param right: The ending index of the range (exclusive, 0-based).
        :return: The result of applying the function to the range.

        The function traverses only the segments that overlap with the given range.
        """
        if not (0 <= left < right <= self.n):
            raise IndexError("Invalid query range")

        # Map range to corresponding leaf nodes in the segment tree
        left += self.n
        right += self.n

        result = self.identity  # Start with the identity value

        # Traverse segments and combine results
        while left < right:
            if left % 2 == 1:  # If left is a right child
                result = self.func(result, self.tree[left])
                left += 1  # Move to the next segment
            if right % 2 == 1:  # If right is a right child
                right -= 1  # Move to the parent node range
                result = self.func(result, self.tree[right])

            # Move to the next level of internal nodes
            left //= 2
            right //= 2

        return result
